pandas/tests/indexes/interval/test_interval.py

```diff
@@ -873,15 +873,17 @@ def test_symmetric_difference(self, closed):
 
     @pytest.mark.parametrize('op_name', [
         'union', 'intersection', 'difference', 'symmetric_difference'])
-    def test_set_operation_errors(self, closed, op_name):
+    def test_set_incompatible_types(self, closed, op_name):
         index = self.create_index(closed=closed)
         set_op = getattr(index, op_name)
 
         # non-IntervalIndex
-        msg = ('the other index needs to be an IntervalIndex too, but '
-               'was type Int64Index')
-        with pytest.raises(TypeError, match=msg):
-            set_op(Index([1, 2, 3]))
+        if op_name == 'difference':
```

Comment:
why is this different?

Reply:
This is due to the changes to the behavior of performing set operations with an interval index and an incompatible type. These operations should no longer raise an error.

Reply:
right as I commented above, i find it strange that *only* difference is an issue here.

Reply:
This is due to the fact that difference operations behave a bit differently in that difference with mismatching types returns the original dtype, which was the case before and discussed in this comment.

https://github.com/pandas-dev/pandas/pull/23538#discussion_r232905638

Reply:
> i find it strange that _only_ difference is an issue here.

iirc it's not so much that `difference` is the only one for which there's an issue, but more so that it's the only one that's handled consistently throughout the codebase in terms of the return type, so it's the only one being special cased here (I could be misremembering).

For some examples with `intersection` (in principle this _could_ always return the same type as the calling index):
```python
In [2]: fidx = pd.Float64Index([0, 1, 1.1, 2.2])

In [3]: iidx = pd.Int64Index([0, 1, 2, 3])

In [4]: dti = pd.date_range('2019-01-01', periods=3)

In [5]: fidx.intersection(iidx)
Out[5]: Int64Index([0, 1], dtype='int64')

In [6]: iidx.intersection(fidx)
Out[6]: Float64Index([0.0, 1.0], dtype='float64')

In [7]: fidx.intersection(dti)
Out[7]: Index([], dtype='object')

In [8]: dti.intersection(fidx)
Out[8]: DatetimeIndex([], dtype='datetime64[ns]', freq=None)
```
`Float64Index` and `Int64Index` seems to be returning the type of `other`.  `DatetimeIndex` looks to always be returning it's own type when it's the caller, but when it's being called by a `Float64Index` or `Int64Index` an `object` dtype `Index` is returned.

Strangely, a `FutureWarning` is given when trying to intersect `DatetimeIndex` and `TimedeltaIndex` saying that a `TypeError` will be raised in the future:
```python-traceback
In [11]: dti.intersection(pd.timedelta_range('0 days', periods=3))
pandas\core\indexes\datetimelike.py:152: FutureWarning: Passing timedelta64-dtype data is deprecated, will raise a TypeError in a future version
pandas\core\indexes\datetimes.py:628: FutureWarning: Passing timedelta64-dtype data is deprecated, will raise a TypeError in a future version
  other = DatetimeIndex(other)
Out[11]: DatetimeIndex([], dtype='datetime64[ns]', freq=None)
```

Reply:
ok that makes sense. @ArtinSarraf can you add a TODO mentioned this (in both the test and here in the code).

Reply:
done